### PR TITLE
[WIP] Coverage analysis support

### PR DESCRIPTION
**Pull request by rafie**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -11,11 +11,17 @@ commands:
       test_params:
         type: string
         default: ""
-    steps: 
+    steps:
       - checkout
       - run:
           name: Submodule checkout
-          command: git submodule update --init --recursive
+          command: |
```
Comment (MeirShpilraien) [inline .circleci/config.yml]: @rafie can you explain what is this for?

File: .circleci/config.yml [hunk 2]
```diff
@@ -188,6 +227,14 @@ workflows:
...
             tags:
               only: /.*/
+      - coverage:
+          requires:
+            - linux-debian
+          filters:
+            branches:
+              only: /.*/
+            tags:
+              only: /.*/
```
Comment (MeirShpilraien) [inline .circleci/config.yml]: How long is it taking, should we run it only on nightly?

File: build/scritps/valgrind.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
```
Comment (MeirShpilraien) [inline build/scritps/valgrind.sh]: What is this script is used for?

File: .circleci/config.yml (hunk 2 above)
Comment (rafie) [reply .circleci/config.yml]: I don't think it should take very long after I fix the mixup with the Valgrind build. It should be possible to run it per-commit.

Comment (MeirShpilraien) [reply .circleci/config.yml]: :+1:

File: .circleci/config.yml (hunk 1 above)
Comment (rafie) [reply .circleci/config.yml]: This is an optimization for the dependency checkout. If we perform coverage analysis following the controlling build, we can save the dependencies in the build and retrieve them in the coverage phase. However, it looks that I might revert it if it turns out that the time gain doesn't worth the complexity.

Comment (MeirShpilraien) [reply .circleci/config.yml]: :+1:

**Comment by MeirShpilraien** [comment]
@rafie Can this be merged?

File: build/scritps/valgrind.sh (hunk 3 above)
Comment (rafie) [reply build/scritps/valgrind.sh]: This is used primarily for Callgrind invocation, and in general in any case you'd like to run valgrind without RLTest.

**Comment by rafie** [comment]
@MeirShpilraien We're still missing some functionality here: 
- [ ] Prevent Valgrind involvement in coverage builds/runs
- [ ] Remove build artifacts caching from .circleci/config.yml

**Comment by MeirShpilraien** [comment]
Use `make GCOV=1` to compile with coverage support and then `make coverage_report` to create the report.

**Pull request closed without merging**